pkg/controller/hibernation/hibernation_controller_test.go

```diff
@@ -289,6 +293,158 @@ func TestReconcile(t *testing.T) {
 
 }
 
+func TestHibernateAfter(t *testing.T) {
+	logger := log.New()
+	logger.SetLevel(log.DebugLevel)
+
+	scheme := runtime.NewScheme()
+	corev1.AddToScheme(scheme)
+	batchv1.AddToScheme(scheme)
+	hivev1.AddToScheme(scheme)
+	hivev1.AddToScheme(scheme)
+
+	cdBuilder := testcd.FullBuilder(namespace, cdName, scheme).Options(
+		func(cd *hivev1.ClusterDeployment) {
+			cd.Spec.Installed = true
+			cd.Status.ClusterVersionStatus = &configv1.ClusterVersionStatus{
+				Desired: configv1.Update{
+					Version: "4.4.9",
+				},
+			}
+		},
+	)
+	o := clusterDeploymentOptions{}
+
+	tests := []struct {
+		name          string
+		setupActuator func(actuator *mock.MockHibernationActuator)
+		cd            *hivev1.ClusterDeployment
+
+		installedTime      time.Time
+		hibernateAfter     string
+		hibernateCondition hivev1.ClusterDeploymentCondition
+		expectRequeueAfter bool
+		expectedPowerState hivev1.ClusterPowerState
+	}{
+		{
+			name: "cluster due for hibernate no condition", // cluster that has never been hibernated and thus has no running condition
+			cd: cdBuilder.GenericOptions(testgeneric.WithAnnotation(constants.HibernateAfterAnnotation, "8h")).Options(
+				testcd.InstalledTimestamp(time.Now().Add(-10 * time.Hour))).Build(),
+			expectedPowerState: hivev1.HibernatingClusterPowerState,
+		},
+		{
+			name: "cluster not yet due for hibernate no condition", // cluster that has never been hibernated and thus has no running condition
+			cd: cdBuilder.GenericOptions(testgeneric.WithAnnotation(constants.HibernateAfterAnnotation, "12h")).Options(
+				testcd.InstalledTimestamp(time.Now().Add(-10 * time.Hour))).Build(),
+			expectRequeueAfter: true,
+			expectedPowerState: "",
+		},
+		{
+			name: "cluster with running condition due for hibernate",
+			cd: cdBuilder.GenericOptions(
+				testgeneric.WithAnnotation(constants.HibernateAfterAnnotation, "8h")).Options(
+				testcd.WithCondition(hibernatingCondition(corev1.ConditionFalse, hivev1.RunningHibernationReason, 9*time.Hour)),
+			).Options(
+				testcd.InstalledTimestamp(time.Now().Add(-10 * time.Hour))).Build(),
+			expectedPowerState: hivev1.HibernatingClusterPowerState,
+		},
+		{
+			name: "cluster with running condition not due for hibernate",
+			cd: cdBuilder.GenericOptions(
+				testgeneric.WithAnnotation(constants.HibernateAfterAnnotation, "20h")).Options(
+				testcd.WithCondition(hibernatingCondition(corev1.ConditionFalse, hivev1.RunningHibernationReason, 6*time.Hour)),
+			).Options(
+				testcd.InstalledTimestamp(time.Now().Add(-10 * time.Hour))).Build(),
+			expectRequeueAfter: true,
+			expectedPowerState: "",
+		},
+		{
+			name: "cluster stuck resuming condition due for hibernate",
+			cd: cdBuilder.GenericOptions(
+				testgeneric.WithAnnotation(constants.HibernateAfterAnnotation, "8h")).Options(
+				// TODO: check false below
+				testcd.WithCondition(hibernatingCondition(corev1.ConditionFalse, hivev1.ResumingHibernationReason, 9*time.Hour)),
```

Comment:
If the cluster is resuming, then the Hibernating condition will have a status of True.

Reply:
These tests were part of where I got into trouble on the first hack day I worked on this. We can't tell if you're stuck resuming or for how long because the hibernating condition stays true, so the transition time is never updated, it's still set to when you went to hibernating. I'm going to remove these tests, getting stuck in one of those phases is outside the scope of this I think.